onnxruntime/core/providers/cpu/math/einsum_utils/einsum_typed_compute_processor.cc

```diff
@@ -0,0 +1,371 @@
+    } else {  // This dimension is not reduced (i.e.) it appears in the output after processing these 2 operands
+      // Both the left and right operands have non-trivial dimension value along this axis
+      // They must be equal
+      if (has_left_dim && has_right_dim) {
+        ORT_ENFORCE(left_dim == right_dim, "Einsum op: Input shapes do not align");
+        lro.push_back(i);
+        lro_size *= left_dim;
+      } else if (has_left_dim) {
+        // The left operand has non-trivial dimension value
+        lo.push_back(i);
+        lo_size *= left_dim;
+      } else {
+        // The right operand may or may not have non-trivial dim value
+        // If it has trivial dim value (1),
+        // it will just form a trailing dimension for the right operand
+        ro.push_back(i);
+        ro_size *= right_dim;
+      }
+    }
+  }
+
+  // Permutate the left operand so that the axes order go like this: [lro, lo, reduce_dims, ro]
+  std::vector<size_t> left_permutation;
+  left_permutation.reserve(lro.size() + lo.size() + reduce_dims.size() + ro.size());
+  left_permutation.insert(left_permutation.end(), lro.begin(), lro.end());
+  left_permutation.insert(left_permutation.end(), lo.begin(), lo.end());
+  left_permutation.insert(left_permutation.end(), reduce_dims.begin(), reduce_dims.end());
+  left_permutation.insert(left_permutation.end(), ro.begin(), ro.end());
+  if (EinsumOp::IsTransposeRequired(current_left ? current_left->Shape().GetDims().size() : left_dims.size(),
+                                    left_permutation)) {
+    current_left = EinsumOp::Transpose(current_left ? *current_left : left,
+                                       current_left ? current_left->Shape().GetDims() : left_dims,
+                                       left_permutation, allocator_, einsum_cuda_assets_,
+                                       device_transpose_func_);
+  }
+
+  // Permutate the right operand so that the axes order go like this: [lro, reduce_dims, ro, lo]
+  std::vector<size_t> right_permutation;
+  right_permutation.reserve(lro.size() + lo.size() + reduce_dims.size() + ro.size());
+  right_permutation.insert(right_permutation.end(), lro.begin(), lro.end());
+  right_permutation.insert(right_permutation.end(), reduce_dims.begin(), reduce_dims.end());
+  right_permutation.insert(right_permutation.end(), ro.begin(), ro.end());
+  right_permutation.insert(right_permutation.end(), lo.begin(), lo.end());
+  if (EinsumOp::IsTransposeRequired(current_right ? current_right->Shape().GetDims().size() : right_dims.size(),
+                                    right_permutation)) {
+    current_right = EinsumOp::Transpose(current_right ? *current_right : right,
+                                        current_right ? current_right->Shape().GetDims() : right_dims,
+                                        right_permutation, this->allocator_, this->einsum_cuda_assets_,
+                                        this->device_transpose_func_);
```

Comment:
nit: 'this->' isn't needed and the trailing '_' in the name implies it's a class member

Reply:
Whoops. Fixed.